### ISSUE TITLE
Question on ApiFilter application to generating reflection info

### ISSUE THREAD

**Issue by Stoffelche** [EWSoftware/SHFB#166]
Hi, I have coded a plug-in which creates an ApiFilter based on the content  in the comment files.
I was wondering why, when I exclude a namespace, then include a class which has nested classes, the nested classes are included as well. For non-nested classes, when I exclude the namespace, the classes are excluded as well, unless explicitly included.
I wanted to find out why nested classes are treated differently, by setting a breakpoint somewhere where the ApiFilter is applied during the generation of the reflection info. 
However I was unable to find the piece of code or debug somehow the process of reflection info generation. I am aware that this is an MSbuild process controlled through a generated proj-file.
But would it be possible to debug this process as well? I imagine there is some assembly of the shfb involved which is loaded during this MS build process?
Thanks, Stefan

**Comment by Stoffelche** [comment]
I had been thinking that all the working code is part of the SandCastleBuilder.sln
In the meanwhile I found the MRefBuilder.csproj, and found that this is where the generation of the reflection file takes place. One crude way to debug it, would be to insert a breakpoint in the Sandcastle build proces at the point where the MRefBuild process is started. 
In the MRefBuilder I put a Thread.Sleep that gives me enough time to attach to the process.
Alternatively I could copy the proj and config that is generated for the process and edit the command line for debugging the MRefBuilder. But this would be tedious, because after a change to the ApiFilter I first need to have the config file regenerated and copy it before the next debug.
Is there another alternative for debugging the various Build processes launched during the SHFB-"Build"?
Thanks, Stefan

**Comment by EWSoftware** [comment]
I'd have to go look to be sure but nested classes are probably treated as children of the parent type if not explicitly included or excluded.  As such, they'd inherit the parent's API visibility.

To debug MRefBuilder, you can open the MRefBuilder project and set its debug properties.  You can either run *MRefBuilder.exe* and specify the command line parameters to tell it what assembly to parse and the configuration file to use or, you can run *MSBuild.exe* and specify the *GenerateRefInfo.proj* from a SHFB build as the parameter.  You can kick off a SHFB build and abort it after the reflection information is generated.  You can then use the working folder and the project from the build as the debug settings for MRefBuilder.  You can also hand edit the configuration files it generates in the working folder to test various changes.

Setting breakpoints in MRefBuilder then works as expected under the above scenarios.  This is true of all the other tools as well.

**Comment by Stoffelche** [comment]
Ok, thanks for this info.
For now I have solved it in my plug-in by explicitly adding the excluded nested classes to the ApiFilter.
BTW I have added support for <includeMembers/> and <includeMembers visibility='public,internal'/>
to my plug-in, which allows for overriding the exclusion-mechanism of the XmlContentFilter.
(visibility attribute optional, it accepts a comma separated list of visibiltiy to include).
Also, I added support for the <exclude/> tag as it does not seem to work by default..
If interested I can send you an update of the sources.
Thanks, Stefan
